qutip/tests/core/test_qobjevo.py

```diff
@@ -42,111 +42,442 @@
+                                         lambda O1, O2: liouvillian(O1, [O2])],
+                             ids=['lindblad', 'lindblad_chi',
+                                  'sprepost', 'liouvillian'])
+    def test_superoperator_qobj(self, form, superop):
+        t = self._rand_t()
+        obj1 = self.qobjevos[form]
+        obj1_t = self.qobjevos[form](t)
+        qobj = rand_herm(self.N)
+
+        as_qevo = superop(obj1, qobj)(t)
+        as_qobj = superop(obj1_t, qobj)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+        as_qevo = superop(qobj, obj1)(t)
+        as_qobj = superop(qobj, obj1_t)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+    @pytest.mark.parametrize('superop', [lindblad_dissipator,
+                                         partial(lindblad_dissipator, chi=0.5),
+                                         sprepost,
+                                         lambda O1, O2: liouvillian(O1, [O2])],
+                             ids=['lindblad', 'lindblad_chi',
+                                  'sprepost', 'liouvillian'])
+    def test_superoperator(self, form, extra_form, superop):
+        t = self._rand_t()
+        obj1 = self.qobjevos[form]
+        obj1_t = self.qobjevos[form](t)
+        obj2 = self.qobjevos[extra_form]
+        obj2_t = self.qobjevos[extra_form](t)
+
+        as_qevo = superop(obj1, obj2)(t)
+        as_qobj = superop(obj1_t, obj2_t)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+    def test_shift(self, form):
+        t = self._rand_t() / 2
+        dt = self._rand_t() / 2
+        obj = self.qobjevos[form]
+        shifted = obj._shift()
+        _assert_qobj_almost_eq(obj(t + dt), shifted(t, args={"_t0": dt}))
+
+    def test_QobjEvo_apply(self):
+        "QobjEvo apply"
+        obj = self.qobjevos["no_args"]
+        transposed = obj.apply(_trans)
+        _assert_qobjevo_equivalent(transposed, obj.trans())
+
+    def test_mul_vec(self, form):
+        "QobjEvo mul_vec"
+        t = self._rand_t()
+        N = self.N
+        vec = np.arange(N)*.5+.5j
+        op = self.qobjevos[form]
+        assert np.allclose(op(t).full() @ vec,
+                           op.mul_vec(t, vec))
+
+    def test_mul_mat(self, form):
+        "QobjEvo mul_mat"
+        t = self._rand_t()
+        N = self.N
+        mat = np.random.rand(N, N) + 1 + 1j * np.random.rand(N, N)
+        matF = np.asfortranarray(mat)
+        op = self.qobjevos[form]
+        Qo1 = op(t)
+        assert np.allclose(Qo1.full() @ mat, op.mul_mat(t, mat))
+        assert np.allclose(Qo1.full() @ matF, op.mul_mat(t, matF))
+
+    def test_expect_psi(self, form):
+        "QobjEvo expect psi"
+        t = self._rand_t()
+        N = self.N
+        vec = _data.dense.fast_from_numpy(np.arange(N)*.5 + .5j)
+        op = self.qobjevos[form]
+        _expect = _data.expect_csr_dense
+        Qo1 = op(t)
+        assert np.allclose(_expect(Qo1.data, vec), op.expect(t, vec))
```

Comment:
all lines with `_expect` in them will clash with the new PR that I'm shortly going to make which puts in the dispatchers properly.  If you're ready to merge this soon, then leave them as they are right now and I'll rebase the new PR on top of this once it's merged.  Otherwise, this is just a warning about why there's a huge merge conflict.

Reply:
When will your PR be ready? I would like to merge this one quickly mostly for the liouvillian fix and auto-compilation of QobjEvo which affect solvers.

Reply:
yeah, merge yours first